sync/blocksync/syncer_test.go

```diff
@@ -0,0 +1,319 @@
+	env := newTestEnvironment(t, 10).withDefaultBlockProvider()
+
+	syncer, err := env.createSyncer(5, 3)
+	require.NoError(t, err)
+	require.NotNil(t, syncer)
+
+	err = syncer.Wait(context.Background())
+	require.ErrorIs(t, err, synccommon.ErrWaitBeforeStart)
+}
+
+func TestBlockSyncer_ParameterizedTests(t *testing.T) {
+	tests := []struct {
+		name                     string
+		numBlocks                int
+		prePopulateBlocks        []int
+		fromHeight               uint64
+		parentsToGet             uint64
+		expectedBlocks           []int
+		verifyZeroBlocksReceived bool
+	}{
+		{
+			name:           "normal case - all blocks retrieved from network",
+			numBlocks:      10,
+			fromHeight:     5,
+			parentsToGet:   3,
+			expectedBlocks: []int{3, 4, 5},
+		},
+		{
+			name:                     "all blocks already available",
+			numBlocks:                10,
+			prePopulateBlocks:        []int{3, 4, 5},
+			fromHeight:               5,
+			parentsToGet:             3,
+			expectedBlocks:           []int{3, 4, 5},
+			verifyZeroBlocksReceived: true,
+		},
+		{
+			name:              "some blocks already available",
+			numBlocks:         10,
+			prePopulateBlocks: []int{4, 5},
+			fromHeight:        5,
+			parentsToGet:      3,
+			expectedBlocks:    []int{3, 4, 5},
+		},
+		{
+			name:              "most recent block missing",
+			numBlocks:         10,
+			prePopulateBlocks: []int{3, 4},
+			fromHeight:        5,
+			parentsToGet:      3,
+			expectedBlocks:    []int{3, 4, 5},
+		},
+		{
+			name:           "edge case - from height 1",
+			numBlocks:      10,
+			fromHeight:     1,
+			parentsToGet:   1,
+			expectedBlocks: []int{1},
+		},
+		{
+			name:           "single block sync",
+			numBlocks:      10,
+			fromHeight:     7,
+			parentsToGet:   1,
+			expectedBlocks: []int{7},
+		},
+		{
+			name:           "large sync - many blocks",
+			numBlocks:      50,
+			fromHeight:     40,
+			parentsToGet:   35,
+			expectedBlocks: []int{6, 10, 20, 30, 40},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			env := newTestEnvironment(t, tt.numBlocks).withDefaultBlockProvider()
+
+			if len(tt.prePopulateBlocks) > 0 {
+				env.prePopulateBlocks(tt.prePopulateBlocks...)
+			}
+
+			syncer, err := env.createSyncer(tt.fromHeight, tt.parentsToGet)
+			require.NoError(t, err)
+
+			ctx := context.Background()
+			require.NoError(t, syncer.Start(ctx))
+			require.NoError(t, syncer.Wait(ctx))
+
+			env.verifyBlocksInDB(t, tt.expectedBlocks)
+
+			if tt.verifyZeroBlocksReceived {
+				// Client should not have received any block requests since all blocks were on disk
+				require.Equal(t, int32(0), env.client.BlocksReceived())
+			}
+		})
+	}
+}
+
+func TestBlockSyncer_ContextCancellation(t *testing.T) {
+	// Allow some time for startup for cancellation
+	env := newTestEnvironment(t, 10).withCustomBlockProvider(func(hash common.Hash, height uint64) *types.Block {
+		time.Sleep(100 * time.Millisecond)
+		return nil
+	})
```

Comment:
The sleep (and custom block provider) don't seem required/relevant for this test.
```suggestion
	env := newTestEnvironment(t, 10).withDefaultBlockProvider()
```
As it is currently written, it seems like a racy test, but it actually isn't. This sleep isn't required.